File: 2022/go/day2/day2.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input, _ := os.ReadFile("../../input/day2.txt")
	lines := strings.Split(string(input), "\n")

	score1 := 0
	for _, line := range lines {
		score1 += game1(line)
	}
	fmt.Printf("part 1: %v\n", score1)

	score2 := 0
	for _, line := range lines {
		score2 += game2(line)
	}
	fmt.Printf("part 2: %v\n", score2)
}

var toWin = []byte{1, 2, 0}  // to win 0 (Rock) you need 1 (Paper) etc
var toLose = []byte{2, 0, 1} // to lose 0 (Rock) you need 2 (Scissors) etc

func game1(line string) int {
	shape1 := line[0] - 'A' // 0=Rock, 1=Paper, 2=Scissors
	shape2 := line[2] - 'X'

	shapeScore := shape2 + 1
	return int(shapeScore) + gameScore(shape1, shape2)
}

func game2(line string) int {
	shape1 := line[0] - 'A'
	shape2 := shape1
	switch line[2] {
	case 'X':
		shape2 = toLose[shape1]
	case 'Z':
		shape2 = toWin[shape1]
	}
	shapeScore := shape2 + 1
	return int(shapeScore) + gameScore(shape1, shape2)
}

func gameScore(shape1 byte, shape2 byte) int {
	if toWin[shape1] == shape2 {
		return 6
	} else if toLose[shape1] == shape2 {
		return 0
	} else {
		return 3
	}
}

File: 2022/java/Day2.java
import java.nio.file.Files;
import java.nio.file.Path;

class Day2 {

    static int[] toWin = { 1, 2, 0 };  //for  0/Rock -> 1/Paper,    1/Paper -> 2/Scissors, 2/Scissors -> 0/Rock
    static int[] toLose = { 2, 0, 1 }; //for  0/Rock -> 2/Scissors, 1/Paper -> 0/Rock    , 2/Scissors -> 1/Paper

    public static void main(String ... args) throws Exception {
        var input = Files.readString(Path.of("../input/day2.txt"));
        var score = input.lines()
            .map(line -> game1(line))
            .mapToInt(game -> score(game)).sum();
        System.out.printf("part 1: %d\n", score);

        var score2 = input.lines()
            .map(line -> game2(line))
            .mapToInt(game -> score(game)).sum();
        System.out.printf("part 2: %d\n", score2);
    }

    static int[] game1(String line) {
        return new int[] { line.charAt(0) - 'A', line.charAt(2)-'X' };
    }

    static int[] game2(String line) {
        int shape1 = line.charAt(0)-'A';
        int shape2 = switch (line.charAt(2)) {
            case 'X' -> toLose[shape1];
            case 'Z' -> toWin[shape1];
            default -> shape1;
        };
        return new int[] { shape1, shape2 };
    }

    static int score(int[] game) {
        var shapeScore = game[1] + 1;
        return shapeScore + gameScore(game[0], game[1]);
    }

    static int gameScore(int shape1, int shape2) {
        if (toWin[shape1] == shape2) return 6;
        if (toLose[shape1] == shape2) return 0;
        else return 3;
    }
}

    